File: java/searchRotatedSortedArray.java
public class searchRotatedSortedArray {
	public int findPivot(int[] nums, int low, int high){
		if(high - low == 1){
			if(nums[low] > nums[high]) 
				return high; 
			else 
				return -1; //no pivot was found, the array is not rotated
		}
		
		int mid = (low + high) / 2;
		if(nums[mid] > nums[high]){
			return findPivot(nums, mid, high);
		}
		else{
			return findPivot(nums, low, mid);
		} 		
	} 

	public int search(int[] nums, int target){
		int n = nums.length;
		int pivot = findPivot(nums,0,n-1);
		System.out.print("Pivot is at: " + pivot + "\n");

		if(pivot != -1){
			if(nums[n-1] >= target){
				return BinarySearch(nums, target, pivot, n-1, true);
			}
			else{
				return BinarySearch(nums, target, 0, pivot-1, false);
			}
		}
		else{
			return BinarySearch(nums, target, 0, n-1, true);
		}		
	}

	public int BinarySearch(int[] nums, int target, int low, int high, boolean increase){
		if(high - low == 1){
			if(nums[low] == target)
				return low;
			else if(nums[high] == target)
				return high;
			else
				System.out.println("Target is not found");
				return -1; //target not found
		}

		int mid = (high + low) / 2;
		if(increase == true){ 
			//increasing order
			if(nums[mid] <= target){
				return BinarySearch(nums, target, mid, high, increase);
			}
			else{
				return BinarySearch(nums, target, low, mid, increase);
			}
		}
		else{
			//decreasing order
			if(nums[mid] <= target){
				return BinarySearch(nums, target, low, mid, increase);
			}
			else{
				return BinarySearch(nums, target, mid, high, increase);
			}
		}
	}

	public static void main (String args[]){
		searchRotatedSortedArray test = new searchRotatedSortedArray();
		int[] numbers = {9,8,7,0,1,2,3,4,5,6};
		int target = 0;
		int res = test.search(numbers, target);
		System.out.println(res);
	}
}

File: python/SearchRotatedSortedArray.py
'''
Created on Oct 18, 2015

@author: mingchen7
'''

import math

class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        if len(nums) == 1 and nums[0] == target:
            return 0
        else:        
            imin,imax = 0, len(nums)-1
            if nums[imin] > nums[imax]: #rotated
                return self.SearchRotated(nums,target,imin,imax)
            else:
                return self.BinarySearch(nums, target, imin, imax)
        
        return -1
    
    def SearchRotated(self,nums,target,imin,imax):
        if imax-imin+1 == 2:
            if nums[imin] == target: return imin
            elif nums[imax] == target: return imax
            else: return -1
        
        mid = int(math.floor((imax+imin)/2))
        if nums[imin] > nums[mid]: # left rotated
            i1 = self.SearchRotated(nums,target,imin,mid)
            i2 = self.BinarySearch(nums,target,mid+1,imax)
        elif nums[mid+1] > nums[imax]: # right rotated
            i1 = self.BinarySearch(nums,target,imin,mid)
            i2 = self.SearchRotated(nums,target,mid+1,imax)
        else:
            i1 = self.BinarySearch(nums, target, imin, mid)
            i2 = self.BinarySearch(nums, target, mid+1, imax)
            
        if i1 != -1: return i1
        elif i2 != -1: return i2
        else: return -1
            
    def BinarySearch(self,nums,target,imin,imax):        
        while imin<=imax:
            imid = int(math.floor((imax+imin)/2))
            if nums[imid] == target:
                return imid
            elif nums[imid] < target:
                imin = imid+1
            else:
                imax = imid-1
            
        return -1
    
s = Solution()
# nums = [8,9,10,0,1,2,3,4,5,6,7]
nums = [1,3]
print s.search(nums,1)

